### PR TITLE
Speed-up warp_affine and fix bugs in RandomAffine [WIP: to add tests]

### PR DESCRIPTION
**Pull request by ducha-aiki**
Current implementation:

> inp=(512, 3, 256, 256), PS=224, dev=cpu, 1.2719683647155762, sec
> inp=(512, 3, 256, 256), PS=32, dev=cpu, 0.03614664077758789, sec
> inp=(256, 1, 64, 64), PS=224, dev=cpu, 0.6061491966247559, sec
> inp=(256, 1, 64, 64), PS=32, dev=cpu, 0.010976791381835938, sec

> inp=(512, 3, 256, 256), PS=224, dev=cuda:0, 0.2985086441040039, sec
> inp=(512, 3, 256, 256), PS=32, dev=cuda:0, 0.00632786750793457, sec
> inp=(256, 1, 64, 64), PS=224, dev=cuda:0, 0.1348881721496582, sec
> inp=(256, 1, 64, 64), PS=32, dev=cuda:0, 0.003765583038330078, sec

Proposed:

> Benchmarking warp_affine
> inp=(512, 3, 256, 256), PS=224, dev=cpu, 0.44802093505859375, sec
> inp=(512, 3, 256, 256), PS=32, dev=cpu, 0.016437292098999023, sec
> inp=(256, 1, 64, 64), PS=224, dev=cpu, 0.07928061485290527, sec
> inp=(256, 1, 64, 64), PS=32, dev=cpu, 0.003212451934814453, sec

> inp=(512, 3, 256, 256), PS=224, dev=cuda:0, 0.26599621772766113, sec
> inp=(512, 3, 256, 256), PS=32, dev=cuda:0, 0.0008182525634765625, sec
> inp=(256, 1, 64, 64), PS=224, dev=cuda:0, 0.0015521049499511719, sec
> inp=(256, 1, 64, 64), PS=32, dev=cuda:0, 0.0006346702575683594, sec
>

### CODE REVIEW

**Comment by ducha-aiki** [comment]
To run speedtest:
make benchmark

**Comment by edgarriba** [comment]
@ducha-aiki is it ready to review?

**Comment by ducha-aiki** [comment]
Not yet, tests are missing.

**Comment by ducha-aiki** [comment]
Now:
image

On master:
image

File: kornia/augmentation/param_gen.py [hunk 1]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
     return params
 
 
+def _compose_affine_matrix_3x3(translations: torch.Tensor,
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: rename to `_get_affine_matrix`  and provide `_get_inverse_affine_matrix`

File: kornia/augmentation/param_gen.py [hunk 2]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+                               center: torch.Tensor,
+                               scale: torch.Tensor,
+                               angle: torch.Tensor,
+                               sx: Optional[torch.Tensor],
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: default value to None ?

File: kornia/augmentation/param_gen.py [hunk 3]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+                               sy: Optional[torch.Tensor]) -> torch.Tensor:
+    r"""Composes affine matrix Bx3x3 from the components
+    Returns:
+        torch.Tensor: params to be passed to the affine transformation.
+    """
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: should we check any shape?

File: kornia/augmentation/param_gen.py [hunk 4]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+        x, y = torch.split(center, 1, dim=-1)
+        x = x.view(-1)
+        y = y.view(-1)
+        sx_tan = torch.tan(sx)  # type: ignore
+        sy_tan = torch.tan(sy)  # type: ignore
+        zeros = torch.zeros_like(sx)  # type: ignore
+        ones = torch.ones_like(sx)  # type: ignore
+        shear_mat = torch.stack([ones,   -sx_tan,                 sx_tan * x,  # type: ignore   # noqa: E241
+                                 -sy_tan, ones + sx_tan * sy_tan, sy_tan * (-sx_tan * x + y)],  # noqa: E241
+                                dim=-1).view(-1, 2, 3)
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: maybe we could provide a functionality like `get_shear_matrid2d` similar to rotation

File: kornia/augmentation/param_gen.py [hunk 5]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+        shear_mat = torch.nn.functional.pad(shear_mat, [0, 0, 0, 1], value=0.)
+        shear_mat[..., -1, -1] += 1.0
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: same here, I see this padding often, we could have something like `convert_matrix2d_to_homogeneous`

File: kornia/augmentation/param_gen.py [hunk 6]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+    transform: torch.Tensor = get_rotation_matrix2d(center, -angle, scale)
+    transform[..., 2] += translations  # tx/ty
+    # pad transform to get Bx3x3
+    transform_h = torch.nn.functional.pad(transform, [0, 0, 0, 1], value=0.)
+    transform_h[..., -1, -1] += 1.0
+    if sx is not None:
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: why the code above if sx is None not happening here ?

File: kornia/augmentation/param_gen.py [hunk 7]
```diff
@@ -258,12 +258,46 @@ def _random_rotation_gen(batch_size: int, degrees: FloatUnionType) -> Dict[str,
...
+        transform_h = torch.bmm(transform_h, shear_mat)
```
Comment (edgarriba) [inline kornia/augmentation/param_gen.py]: just curious, why not `torch.matmul` or `@`

Comment (ducha-aiki) [reply kornia/augmentation/param_gen.py]: Just habit :) I associate matmul with mm, i.e. non-batched and @ didn't work in python 2))

File: kornia/augmentation/param_gen.py (hunk 5 above)
Comment (ducha-aiki) [reply kornia/augmentation/param_gen.py]: Good idea. I created `convert_affinematrix_to_homography`

File: kornia/augmentation/param_gen.py (hunk 4 above)
Comment (ducha-aiki) [reply kornia/augmentation/param_gen.py]: Good idea, but I am not sure how to handle translation there. We have a special case here for being around center. The general case is not clear to me...

Comment (shijianjian) [reply kornia/augmentation/param_gen.py:286]: image
@ducha-aiki I am trying to implement the 3D version but I cannot reproduce this translation here. Can you let me know the exact math you did?

Comment (ducha-aiki) [reply kornia/augmentation/param_gen.py:286]: @shijianjian you are right, good catch. I messed up x and y somehow
